### PR TITLE
net/tcp: tcp_sendfile need restore the file location at the end

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
quote from https://man7.org/linux/man-pages/man2/sendfile.2.html:
If offset is not NULL, then it points to a variable holding the
file offset from which sendfile() will start reading data from
in_fd.  When sendfile() returns, this variable will be set to the
offset of the byte following the last byte that was read.  If
offset is not NULL, then sendfile() does not modify the file
offset of in_fd; otherwise the file offset is adjusted to reflect
the number of bytes read from in_fd.
If offset is NULL, then data will be read from in_fd starting at
the file offset, and the file offset will be updated by the call.

The change also align with the implementation at:
libs/libc/misc/lib_sendfile.c

## Impact
Make sendfile more confirm to the standard

### CODE REVIEW

File: net/tcp/tcp_sendfile.c [hunk 1]
```diff
@@ -304,15 +303,6 @@ static uint16_t sendfile_eventhandler(FAR struct net_driver_s *dev,
...
 
-          ret = file_seek(pstate->snd_file,
-                          pstate->snd_foffset + pstate->snd_sent, SEEK_SET);
-          if (ret < 0)
-            {
-              nerr("ERROR: Failed to lseek: %d\n", ret);
-              pstate->snd_sent = ret;
-              goto end_wait;
-            }
-
```
Comment (yamt) [inline net/tcp/tcp_sendfile.c:315]: i guess this breaks retransmission.

**Review by yamt — CHANGES_REQUESTED** [review]

**Comment by yamt** [comment]
btw, file i/o with net_lock held looks quite dangerous to me.
do we have a defined locking order between net_lock and filesystem locks?

File: net/tcp/tcp_sendfile.c (hunk 1 above)
Comment (xiaoxiang781216) [reply net/tcp/tcp_sendfile.c]: Yes, you are right. Please review the new version again.

**Comment by xiaoxiang781216** [comment]
> btw, file i/o with net_lock held looks quite dangerous to me.
> do we have a defined locking order between net_lock and filesystem locks?

All file i/o either doesn't hold net_lock, or temporarily net_unlock. Could you point out the place I missed?

**Comment by yamt** [comment]
> > btw, file i/o with net_lock held looks quite dangerous to me.
> > do we have a defined locking order between net_lock and filesystem locks?
> 
> All file i/o either doesn't hold net_lock, or temporarily net_unlock. Could you point out the place I missed?

isn't sendfile_eventhandler called with the lock held?

**Comment by xiaoxiang781216** [comment]
Oh, yes I look at the wrong location.

**Comment by xiaoxiang781216** [comment]
@yamt I open a new issue(https://github.com/apache/incubator-nuttx/issues/4122) to track this problem. The problem exist before this patch, let's fix it by a new patch?

**Comment by yamt** [comment]
> @yamt I open a new issue(#4122) to track this problem. The problem exist before this patch, let's fix it by a new patch?

ok

**Review by yamt — APPROVED** [review]
lgtm

**Pull request merged**